### ISSUE TITLE
Failsafe state system

### ISSUE THREAD

**Issue by peabody124** [TauLabs/TauLabs#1410]
With most platforms having baro's currently, it might be possible to have a safer
failsafe than motors immediately off, at least to mask brief signal interference.

For example when signal is lost to hold altitude for several seconds before
falling. Possibly even then a period of rapid (powered) descent first to get it 
closer to the ground without hitting terminal velocity -- although this might
still not be safe since the risk of drifting during the fall is very real and takes
the UAV outside of where the pilot intended.

Finally we might want to have the option for RTH, although this will require
much more testing and having large numbers of users reporting that we
have a robust configuration for RTH and good defaults. This will be a more
extensive task.

Either way, this will require the development of a state system to handle the
failsafe behavior more completely. I'd also really like to solicit input with
regards to safety.

One clear advantage of baro as opposed to throttle setpoints is we can at
least guarantee the UAV is descending or holding steady.

**Comment by UKTamo** [comment]
Definitely would like to be able to configure what Failsafe actually does. I for one am not a fan of the drop out the sky methodology of failsafe (have you ever seen a 680 hex drop out the sky with no motors running, it's scary) PH for 10 seconds then RTH would be ideal for GPS enabled craft, or just PH/Alt Hold for 10 seconds before controlled descent could be better than falling out the sky.

**Comment by peabody124** [comment]
One reason we haven't even considered enabling RTH by default is that not enough people have tested it (when intentional) on enough aircrafts. When we have some confidence that everything works all the time, it will be an option. However, making it an option prematurely is extremely dangerous. If it tries to fly away and the operator has no control it can end up anywhere and is much more dangerous than falling out of the sky.

Falling out of the sky is scary for sure, and bad for a vehicle. However, people should always fly responsibly and never over anything or anyone that can be damaged. This is responsible piloting - there are any number of reasons that things can fall out of the sky that RTH would be unable to recover from.

**Comment by brat002** [comment]
May be you can create branch with enabled RTH/Landing for testing? I have expensive quadcopter and want to migrate to Quanton+Taulabs code, but i am afraid to fall from the sky when rc signal will be lost. Thousands $ can be lost.

**Comment by osergios** [comment]
I cannot agree more with UkTamo, at least you could initiate a controlled descent in the event of a failsafe and not a plain drop .

**Comment by brat002** [comment]
Guys, any progress? It`s not last feature for safe flying.

**Comment by robearm** [comment]
Yes I for one have been testing PH and RTH this would be nice to set for fail safe and auto land for GPS failure or low sat lock. I agree setting this feature for failsafe without testing it is crazy but allowing the setting is a must. What type of info is needed to push this forward. The latest build as of July 27 2015 still works other than a change in landing. Landing is very fast and stops short of the ground and hovers. And the landing won't engage until throttle is above 50%.
